finality-aleph/src/network.rs

```diff
@@ -141,191 +222,253 @@ enum SessionStatus {
     //TODO: some optimizations can be made by changing Mutex to RwLock
     network: N,
     protocol: Cow<'static, str>,
 
     /// Outgoing events to the consumer.
-    sessions: Arc<Mutex<HashMap<SessionId, SessionData>>>,
+    sessions: Arc<Mutex<HashMap<SessionId, SessionData<D>>>>,
 
-    net_command_tx: mpsc::UnboundedSender<SessionCommand>,
-    net_command_rx: mpsc::UnboundedReceiver<SessionCommand>,
+    commands_for_session: mpsc::UnboundedSender<SessionCommand<D>>,
+    commands_from_rush: mpsc::UnboundedReceiver<SessionCommand<D>>,
 
     peers: Arc<Mutex<Peers>>,
-    phantom: PhantomData<B>,
+    _phantom: PhantomData<B>,
 }
 
-pub(crate) struct SessionManagar {
-    sessions: Arc<Mutex<HashMap<SessionId, SessionData>>>,
-    net_command_tx: mpsc::UnboundedSender<SessionCommand>,
+pub(crate) struct SessionManager<D: Clone + Encode + Decode> {
+    peer_id: PeerId,
+    sessions: Arc<Mutex<HashMap<SessionId, SessionData<D>>>>,
+    commands_for_session: mpsc::UnboundedSender<SessionCommand<D>>,
 }
 
-impl SessionManagar {
-    // NOTE: later this will also need access to KeyStore :/ (for Reliable Broadcast)
+impl<D: Clone + Encode + Decode> SessionManager<D> {
     pub(crate) fn start_session(
         &self,
         session_id: SessionId,
-        _authorities: Vec<AuthorityId>,
-    ) -> RushNetwork {
-        let (net_event_tx, net_event_rx) = mpsc::unbounded();
+        keychain: KeyBox,
+    ) -> GenericNetwork<D> {
+        let (data_for_rush, data_from_network) = mpsc::unbounded();
+        let auth_data = AuthData {
+            session_id,
+            peer_id: self.peer_id.clone(),
+            node_id: keychain.index(),
+        };
+        let signature = keychain.sign(&auth_data.encode());
         let session_data = SessionData {
-            net_event_tx,
+            data_for_rush,
             status: SessionStatus::InProgress,
-            _authorities,
+            keychain,
+            auth_data: auth_data.clone(),
+            auth_signature: signature.clone(),
         };
         self.sessions.lock().insert(session_id, session_data);
-        RushNetwork {
+        if let Err(e) = self
+            .commands_for_session
+            .unbounded_send(SessionCommand::Meta(auth_data, signature, None))
+        {
+            log::error!(target: "afa", "sending auth command failed in new session: {}", e);
+        }
+        GenericNetwork {
             session_id,
-            net_event_rx,
-            net_command_tx: self.net_command_tx.clone(),
+            data_from_network,
+            commands_for_session: self.commands_for_session.clone(),
         }
     }
 }
 
-impl<B: BlockT + 'static, N: Network<B> + Clone> ConsensusNetwork<B, N> {
+impl<D: Clone + Encode + Decode, B: BlockT + 'static, N: Network<B> + Clone>
+    ConsensusNetwork<D, B, N>
+{
     /// Create a new instance.
     pub(crate) fn new(network: N, protocol: impl Into<Cow<'static, str>>) -> Self {
```

Comment:
`protocol` should be just a `Cow` :cow:

Reply:
Heh, definitely, not my fault though I did not touch this specific line. ;p Changed.